### PR TITLE
BuildInfo plugin for build-data and Docker images

### PR DESCRIPTION
**Pull request by mathieucarbou**

### CODE REVIEW

File: src/main/java/org/terracotta/build/plugins/docker/DockerBuildPlugin.java [hunk 1]
```diff
@@ -205,6 +206,11 @@ private void configureBuildDefaults(Project project, DockerBuildExtension docker
     dockerBuild.getMetadata().put("gradle.build.host", getLocalHostName(project));
     dockerBuild.getMetadata().put("gradle.build.dir", project.getRootDir().getAbsolutePath());
     dockerBuild.getMetadata().put("gradle.build.project", project.getPath());
+
+    BuildInfo buildInfo = BuildInfo.getOrInstall(project);
```
Comment (chrisdennis) [inline src/main/java/org/terracotta/build/plugins/docker/DockerBuildPlugin.java:210]: It feels like this should be being retrieved through the project extension to decouple the way the BuildDataPlugin is implemented from how things are consumed.

File: build.gradle [hunk 2]
```diff
@@ -72,5 +72,9 @@ gradlePlugin {
       id = 'org.terracotta.build.voltron'
       implementationClass = 'org.terracotta.build.plugins.VoltronPlugin'
     }
+    buildInfo {
+      id = 'org.terracotta.build.buildInfo'
```
Comment (chrisdennis) [inline build.gradle:76]: plugin-ids are normally snake-case.

File: src/main/java/org/terracotta/build/plugins/CopyrightPlugin.java [hunk 3]
```diff
@@ -79,6 +75,7 @@ public class CopyrightPlugin implements Plugin<Project> {
 
   @Override
   public void apply(Project project) {
+    Git.getOrInstall(project);
```
Comment (chrisdennis) [inline src/main/java/org/terracotta/build/plugins/CopyrightPlugin.java:78]: Why is this needed? Shouldn't this just pickup the service reference if it's there?

File: src/main/java/org/terracotta/build/services/BuildInfo.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    this.commitHash = git.getCommitHash();
+    this.hasLocalChange = git.hasLocalChange();
+
+    this.version = StructuredVersion.parse(overrideVersion != null ?
+        overrideVersion :
+        (defaultVersion != null ? defaultVersion : projectVersion));
+
+    this.revision = !this.hasLocalChange ?
+        this.commitHash :
+        (this.commitHash + "+" + System.getProperty("user.name") + ":" + git.hash(git.diff(this.commitHash)));
```
Comment (chrisdennis) [inline src/main/java/org/terracotta/build/services/BuildInfo.java:52]: I'm thinking we might want to nail down some of the presentation aspects of `git diff` here to try and standardize things?

File: src/main/java/org/terracotta/build/services/Git.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    return execute(spec -> {
+      spec.args("hash-object", "--stdin");
+      spec.setStandardInput(new ByteArrayInputStream(data.getBytes(StandardCharsets.UTF_8)));
+    });
+  }
+
+  public String execute(Action<ExecSpec> action) {
+    return ExecUtils.execute(getExecOperations(), composite(spec -> {
+      spec.executable(getParameters().getGitExecutable().get());
+      spec.args("--no-pager");
```
Comment (chrisdennis) [inline src/main/java/org/terracotta/build/services/Git.java:60]: It feels like `--no-pager` shouldn't be needed because this should be a non-interactive shell?

**Review by chrisdennis — CHANGES_REQUESTED** [review]

File: src/main/java/org/terracotta/build/services/Git.java (hunk 5 above)
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/services/Git.java:60]: No this is required, at least with the diff command, otherwise it gives en empty output stream.

File: src/main/java/org/terracotta/build/services/BuildInfo.java (hunk 4 above)
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/services/BuildInfo.java:52]: what do you mean ? I don't understand... You mean the different options git diff has to generate a diff ?

File: src/main/java/org/terracotta/build/plugins/CopyrightPlugin.java (hunk 3 above)
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/plugins/CopyrightPlugin.java:78]: My understanding of the `@ServiceReference` Javadoc is that it will pick up the shared service found with the given name, not create it. Besides, the javadoc of the annotation explains what happens if the shared service is not found.

File: src/main/java/org/terracotta/build/services/BuildInfo.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+import static java.lang.Integer.parseInt;
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.joining;
+
+public abstract class BuildInfo implements BuildService<BuildInfo.Parameters> {
```
[the remark this answers is not in the document]
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/services/BuildInfo.java:20]: For 1)

I didn't want a user extension which can be configured: I wanted a service that provides information about the build in the immutable manner.

I went to the fastest route to expose directly the BuildInfo object instead of making an interface, but I can make an interface.

For 2)

We do not force that: the choice of the scope if decided how the shared service is constructed and used. I just provided a singleton method to get a default one.

```
  public static BuildInfo getOrInstall(Project project) {
    return project.getGradle().getSharedServices().registerIfAbsent("buildInfo", BuildInfo.class, spec -> spec.parameters(parameters -> {
      parameters.getDefaultProjectVersion().set(project.findProperty("defaultVersion"));
      parameters.getOverrideProjectVersion().set(project.findProperty("overrideVersion"));
      parameters.getProjectVersion().set(project.getVersion());
      parameters.getGit().set(Git.getOrInstall(project));
    })).get();
  }
```

But nothing prevents another plugin or task to use a newly created service that would be constructed like this:

```
return project.getGradle().getSharedServices().registerIfAbsent("myOtherPorjectBuildInfo", BuildInfo.class, spec ->
```

File: src/main/java/org/terracotta/build/services/BuildInfo.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    this.revision = !this.hasLocalChange ?
+        this.commitHash :
+        (this.commitHash + "+" + System.getProperty("user.name") + ":" + git.hash(git.diff(this.commitHash)));
+  }
+
+  public boolean hasLocalChange() {
+    return hasLocalChange;
+  }
+
+  public String getRevision() {
```
[the remark this answers is not in the document]
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/services/BuildInfo.java:59]: yes, wil fix.

File: build.gradle (hunk 2 above)
Comment (mathieucarbou) [reply build.gradle:76]: will fix

File: src/main/java/org/terracotta/build/services/Git.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  public boolean hasLocalChange() {
+    return !execute(spec -> spec.args("status", "--porcelain=2")).isEmpty();
+  }
+
+  public String getCommitHash() {
+    return Optional.ofNullable(System.getenv("GIT_COMMIT"))
+        .orElseGet(() -> execute(spec -> spec.args("rev-parse", "HEAD")));
+  }
+
+  public String getBranch() {
```
[the remark this answers is not in the document]
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/services/Git.java:41]: will fix. 
so right now some plugins require git or other things, but you are right, this would break `gw assemble`

File: src/main/java/org/terracotta/build/services/Git.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+import static java.util.Arrays.asList;
+import static org.gradle.internal.Actions.composite;
+
+public abstract class Git implements BuildService<Git.Parameters> {
+  public interface Parameters extends BuildServiceParameters {
+    Property<String> getGitExecutable();
+  }
+
+  @Inject
+  public abstract ExecOperations getExecOperations();
```
[the remark this answers is not in the document]
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/services/Git.java:30]: I can add it but what's the point ? According to the javadoc (ProcessForkOptions) this is the project working dir.
And we never configured that for the copyright plugin or in the build-data script, which had the same behaviour.
It think the day we need we can add, but I think also that this day might never come.

File: src/main/java/org/terracotta/build/plugins/docker/DockerBuildPlugin.java (hunk 1 above)
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/plugins/docker/DockerBuildPlugin.java:210]: I will ping you today for more explanation because you are diving too deep in Gradle for me there :P

File: src/main/java/org/terracotta/build/services/Git.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/services/Git.java:30]: I will add the option, set to project root dir. This will clear any question.

File: src/main/java/org/terracotta/build/plugins/BuildInfoPlugin.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import org.terracotta.build.services.Git;
+
+import java.time.Instant;
+import java.util.Objects;
+
+public class BuildInfoPlugin implements Plugin<Project> {
+
+  @Override
+  public void apply(@NotNull Project project) {
+    final Provider<Object> overrideVersion = project.provider(() -> project.findProperty("overrideVersion"));
```
Comment (chrisdennis) [inline src/main/java/org/terracotta/build/plugins/BuildInfoPlugin.java:17]: There are methods for getting property providers:
```java
final Provider<String> overrideVersion = project.getProviders().gradleProperty("overrideVersion");
```
I think those will work here.

File: src/main/java/org/terracotta/build/plugins/docker/DockerBuildPlugin.java [hunk 11]
```diff
@@ -45,7 +47,9 @@ public void apply(Project project) {
 
     DockerBuildExtension buildExtension = ((ExtensionAware) dockerExtension).getExtensions().create("build", DockerBuildExtension.class);
 
-    configureBuildDefaults(project, buildExtension);
+    BuildInfoExtension buildInfoExtension = project.getExtensions().getByType(BuildInfoExtension.class);
```
Comment (chrisdennis) [inline src/main/java/org/terracotta/build/plugins/docker/DockerBuildPlugin.java:50]: If we're going to unconditionally use the extension here we should probably have the `DockerBuildPlugin` apply the `BuildInfoPlugin`.

**Review by chrisdennis — CHANGES_REQUESTED** [review]

File: src/main/java/org/terracotta/build/plugins/docker/DockerBuildPlugin.java (hunk 1 above)
Comment (mathieucarbou) [reply src/main/java/org/terracotta/build/plugins/docker/DockerBuildPlugin.java]: outdated
